qiskit_machine_learning/neural_networks/effective_dimension.py

```diff
@@ -0,0 +1,354 @@
+"""An implementation of the effective dimension algorithm."""
+
+import time
+from typing import Optional, Union, List, Callable, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+
+from qiskit.utils import algorithm_globals
+from qiskit_machine_learning import QiskitMachineLearningError
+
+from .opflow_qnn import OpflowQNN
+from .neural_network import NeuralNetwork
+
+
+class EffectiveDimension:
+    """
+    This class computes the global effective dimension for Qiskit
+    :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`
+    following the definition used in [1].
+
+        **References**
+        [1]: Abbas et al., The power of quantum neural networks.
+        `The power of QNNs <https://arxiv.org/pdf/2011.00027.pdf>`__.
+    """
+
+    def __init__(
+        self,
+        qnn: NeuralNetwork,
+        weight_samples: Union[List[float], np.ndarray, int] = 1,
+        input_samples: Union[List[float], np.ndarray, int] = 1,
+        callback: Optional[Callable[[int, float, float], None]] = None,
+    ) -> None:
+
+        """
+        Args:
+            qnn: A Qiskit :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`,
+                with a specific dimension ``(num_weights)`` that will determine the shape of the
+                Fisher Information Matrix ``(num_input_samples * num_weight_samples, num_weights,
+                num_weights)`` used to compute the global effective dimension for a set of
+                ``input_samples``, of shape ``(num_input_samples, qnn_input_size)``, and
+                ``weight_samples``, of shape ``(num_weight_samples, num_weights)``.
+            weight_samples: An array of neural network parameters (weights), of shape
+                ``(num_weight_samples, num_weights)``, or an ``int`` to indicate the number of
+                parameter sets to sample randomly from a uniform distribution. By default,
+                ``weight_samples = 1``.
+            input_samples: An array of samples to the neural network, of shape
+                ``(num_input_samples, qnn_input_size)``, or an ``int`` to indicate the number of
+                input sets to sample randomly from a normal distribution. By default,
+                ``input_samples = 1``.
+            callback: A callback function for the Monte Carlo sampling.
+        """
+
+        # Store arguments
+        self._weight_samples = None
+        self._input_samples = None
+        self._num_weight_samples = 1
+        self._num_input_samples = 1
+        self._model = qnn
+        self._callback = callback
+
+        # Define weight samples and input samples
+        self.weight_samples = weight_samples  # type: ignore
+        # input setter uses self._model
+        self.input_samples = input_samples  # type: ignore
+
+    def num_weights(self) -> int:
+        """Returns the dimension of the model according to the definition from [1]."""
+        return self._model.num_weights
```

Comment:
`Returns the dimension of the model...`  So far we have given a NeuralNetwork as a qnn into the constructor. Yes it was stored as _model but somehow I imagine the term model to be something I have to read the reference and its not something we refer qnn to be in the constructor docstring. This is also call num_weights not dim or dimension. I think maybe, something more like this perhaps
`Returns the number of weights of the supplied neural network. This is the dimension of the model according to the definition from [1].` where we explain it in terms of the code here and state the correspondence to the paper rather than the latter that being the main thing.  Since we passed in qnn and ought to know num_weights from that, this indirect access to it - is it really useful?

Reply:
I believe `num_weights` can be easily retrieved from qnn, so I don't see a real need for this method.